spec/abstractops.html

```diff
@@ -1428,34 +1428,24 @@ <h1>ParseTemporalTimeZoneString ( _isoString_ )</h1>
     </p>
     <emu-alg>
       1. Assert: Type(_isoString_) is String.
-      1. If _isoString_ does not satisfy the syntax of a |TemporalTimeZoneString| (see <emu-xref href="#sec-temporal-iso8601grammar"></emu-xref>), then
+      1. Let _parseResult_ be ParseText(! StringToCodePoints(_isoString_), |TemporalTimeZoneString|).
+      1. If _parseResult_ is a List of errors, then
         1. Throw a *RangeError* exception.
-      1. Let _z_, _sign_, _hours_, _minutes_, _seconds_, _fraction_ and _name_ be the parts of _isoString_ produced respectively by the |UTCDesignator|, |TimeZoneUTCOffsetSign|, |TimeZoneUTCOffsetHour|, |TimeZoneUTCOffsetMinute|, |TimeZoneUTCOffsetSecond|, |TimeZoneUTCOffsetFractionalPart|, and |TimeZoneIANAName| productions, or *undefined* if not present.
-      1. If _z_ is not *undefined*, then
+      1. Let each of _z_, _offsetString_, and _name_ be the source text matched by the |UTCDesignator|, |TimeZoneNumericUTCOffset|, and |TimeZoneIANAName| Parse Node enclosed by _parseResult_, or an empty sequence of code points if not present.
```

Comment:
Thanks for the detailed review, I'm just going through to make sure that these things are captured in issues. I think they are out of scope for this PR.

> * "Z" and "z" are valid expansions of TimeZoneIANAName (relevant for e.g. `Temporal.TimeZone.from`).

I don't think this is a problem — `Temporal.TimeZone.from('Z')` will fail because there's no such IANA named time zone, just like `Temporal.TimeZone.from('Y')`. Ditto for `Temporal.ZonedDateTime.from('2022-01-11T19:43Z[Z]')`.
 
> * "Etc/GMT+_hour_" strings are _not_ matched by TimeZoneIANAName while "Etc/GMT-_hour_" strings are, despite both being in the IANA Time Zone Database.
> 
> * TimeZoneBracketedAnnotation matches "Etc/GMT±0_digit_" strings, which are not in the IANA Time Zone Database (the current range is actually asymmetric—Etc/GMT+1 to Etc/GMT+12 and Etc/GMT-1 to Etc/GMT-14 plus Etc/GMT±0, but I don't see a problem with requiring Temporal implementations to support everything up to Etc/GMT±23).
> 
> * The result of the preceding two points is ambiguous parsing for "Etc/GMT±0_digit_" strings from TimeZoneBracketedName. Their special-case treatment should probably be either removed or moved into TimeZoneIANAName with its existing TimeZoneIANANameTail production prohibited to match e.g. TimeZoneGMTName.

The above points are covered in #1993, I think.

> * There also seems to be ambiguous parsing for "…±offset[±offset]" strings from TimeZone (matched by both TimeZoneOffsetRequired and TimeZoneNameRequired), although I'm not 100% certain that's a problem.

I don't really think this is a problem either, but for the sake of caution we could fix it by rewriting _TimeZone_:

_TimeZone_ :
-  _TimeZoneUTCOffset_ _TimeZoneBracketedAnnotation_(opt)
-  _TimeZoneBracketedAnnotation_

> * The whole tree is so deep and some productions are so long that it's quite difficult to comprehend. I think it could be improved by condensing productions, in some cases by parameterization like replacing `TimeZoneNumericUTCOffset` with `TimeZoneUTCOffset[+onlyNumeric]` and `ASCIISign` with `Sign[+onlyASCII]` and in others with early errors like "if the source text matched by {TimeZoneIANANameComponent,CalendarNameComponent,FractionalPart} contains more than <N> code points".

I think this is covered in #1984.

> * What is the reason for distinguishing TimeZoneNumericUTCOffset and TimeZoneUTCOffsetName and their subproductions, all of which seem to match exactly the same input?

This is to avoid ambiguity in strings like `1976-11-18T15:23:30+01:00[+01:00]` where TimeZoneNumericUTCOffset matches the first `+01:00` and TimeZoneUTCOffsetName matches the second `+01:00`. Otherwise the code using these productions in ParseZonedDateTimeString can't refer unambiguously to the correct parse node.
 
> * Relatedly, how does this code handle an expansion of TimeZone that includes both TimeZoneNumericUTCOffset and TimeZoneUTCOffsetName (the former through TimeZoneUTCOffset and the latter through TimeZoneBracketedAnnotation)? It seems to discard the TimeZoneUTCOffsetName data.

It depends on which `from()` function is parsing the string. Off the top of my head — in the case of Instant, the bracketed annotation is discarded. In the case of ZonedDateTime and TimeZone, the numeric offset is discarded (in ZonedDateTime, only after verifying that they match). In other cases, both are discarded.

Reply:
> Otherwise the code using these productions in ParseZonedDateTimeString can't refer unambiguously to the correct parse node.

That should be possible without introducing distinct nonterminals, see [The `if` Statement](https://tc39.es/ecma262/multipage/ecmascript-language-statements-and-declarations.html#sec-if-statement-runtime-semantics-evaluation) for an example ("the first _Statement_" vs. "the second _Statement_").

Reply:
Including if either of them are optional? (so "first" and "second" are not always the case)

Reply:
Yes, cf. https://tc39.es/ecma262/multipage/ecmascript-language-statements-and-declarations.html#sec-runtime-semantics-forloopevaluation

Reply:
I've captured this as #2006 to address later. Mind if I go on and merge this?